pkg/controller/build/build_controller.go

```diff
@@ -162,6 +163,10 @@ func (r *ReconcileBuild) Reconcile(request reconcile.Request) (reconcile.Result,
 		ctxlog.Error(ctx, err, "failed to update the Build status", "Build", b.Name)
 		return reconcile.Result{}, err
 	}
+
+	// Increase Build count in metrics
+	buildmetrics.BuildCountInc(b.Spec.StrategyRef.Name)
```

Comment:
This increments the metric for each successful reconcile of a build and not only on its creation. Failed registers are not counted. My gut feeling is that the creation predicate function is the only place where one can be 100 % sure that one currently handles a creation (see https://github.com/SaschaSchwarze0/build/blob/985d2716a0f38ce83885751333acae9c5273bc2f/pkg/controller/buildrun/buildrun_controller.go#L80-L83). I checked whether side effects in predicate functions are a bad practice but did not find something.

Reply:
predicate is better to cover all succeed and fail case.

But for the build, there is also another problem here.

If the controller is restarted, the count will be clean to zero, and `predicate` way will start count the new created build or buildrun from zero.

But count in the reconciliation logic is different. After you start a new controller pod, ALL builds will go through the reconcile logic again. It mean, you can still get the total succeed build count after the pod restart.

So I think the best way like what tekton does, by using the `lister` to query the real total or running taskrun correctly, it won't missed anyone.

both current `succeed` and `predicate` ways cannot cover all cases, that is why I suggest we can have a basic metrics first then continue improving our logic or metric together.